Parameters as a map

clalancette opened the pull request:
@Lobotuerk FYI, here is an alternative proposal to your #3.  I basically took your code and refactored it a bit to clean it up and get rid of some of the hard-coded constants.  Take a look and tell me what you think about this.  @hidmic, any thoughts on this are also welcome.

I should mention that this is really only compile tested; this would need additional testing before we merged it.

Lobotuerk committed:
Fixed duration of callbacks for the timeouts

Lobotuerk committed:
Fixed duration of callbacks for the timeouts

Lobotuerk committed:
kind of dictionarie working

Lobotuerk committed:
Made it work with key_name

Lobotuerk committed:
Keyless dictionary working everything except fast adding and removing

Lobotuerk committed:
Everything but removing parameters is working, set a subs topic to /none or something like that for now

Lobotuerk committed:
Merge branch 'ros2-devel' into old_state

Lobotuerk committed:
Delete cmd_vel_mux_params_test.yaml

Lobotuerk committed:
Merge pull request #1 from Lobotuerk/old_state

Old state with key_name property to plan dict with change on the fly working

Lobotuerk committed:
Merge branch 'ros2-devel' into ros2-devel

Lobotuerk committed:
Last commit of fixes, starting refactor to full map way

Lobotuerk committed:
Last commit of fixes, starting refactor to full map way

Lobotuerk committed:
WORKING IM SO HAPPY

clalancette committed:
More fixes to the map implementation.

This should make it more robust and simplify it.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

Lobotuerk commented:
If its working, I've got no complain. Looks cleaner and more robust. Also, if not tested, #3 was thoroughly tested, so, if for some reason you need to, use it until I test #4 tomorrow morning

Lobotuerk commented:
1- Verification for not having repeated priorities, not working.
2- Adding or removing does not work either, for adding, you can only add one parameter at the time, so the set always returns, you are missing the other 2. For removing, the addInputToParameterMap, does not handle PARAMETER_NOT_SET, so it always returns "invalid parameter"

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -30,37 +30,62 @@
 {
-  std::vector<std::string> names;
-  if (!get_parameter("subscribers.name", names))
+std::vector<std::string> stringSplit(const std::string &str,
```
@Lobotuerk @clalancette consider using `rcpputils::split` instead.

clalancette replied:
I didn't know about that one.  Good call, fixed in 6091ab8532173b77448379318c2d7c1bb47dec2e

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,79 +104,109 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
+      RCLCPP_WARN(get_logger(), "Empty topic for '%s'", parameter.first.c_str());
+      return false;
+    }
+    else if (parameter.second.timeout < 0.0)
```
@clalancette @Lobotuerk no need for `else` if the previous `if` clause does `return`.

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,79 +104,109 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
 {
-  std::vector<std::shared_ptr<CmdVelSub>> new_list(names.size());
-  for (unsigned int i = 0; i < names.size(); i++)
+  std::map<std::string, ParameterValues>::const_iterator parameter;
```
@clalancette @Lobotuerk unused?

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,79 +104,109 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
-                                [&new_sub_name](const std::shared_ptr<CmdVelSub>& sub)
-                                                {return sub->name_ == new_sub_name;});
-    if (old_sub != list_.end())
+    std::string key = parameter.first;
```
@clalancette @Lobotuerk here and elsewhere, use const references to alias `std::pair` attributes.

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,79 +104,109 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
+    std::string key = parameter.first;
+    ParameterValues parameter_values = parameter.second;
+    // Check if parameter subscriber has all its necessary values
+    if (map_.find(key) != map_.cend())
```
@clalancette @Lobotuerk on a second look, `map_.count(key) != 0` is much clearer.

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,79 +104,109 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
 
-    // Fill attributes with a YAML node content
     double new_timeout;
     std::string new_topic;
```
@clalancette @Lobotuerk why do we need auxiliary variables? Why not using `parameter` attributes directly?

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,79 +104,109 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
+  for (std::pair<const std::string, std::shared_ptr<CmdVelSub>> & m : map_)
   {
-    if (!list_[i]->sub_)
+    std::string key = m.first;
```
@clalancette @Lobotuerk same as above.

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,79 +104,109 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
   {
-    if (!list_[i]->sub_)
+    std::string key = m.first;
+    if (!m.second->sub_)
```
@clalancette I'm inclined towards aliasing instead of using `std::pair` attributes. It does not help readability.

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -166,27 +221,120 @@ void CmdVelMux::configureFromParameters(const std::vector<std::string> & names,
 }
 
-void CmdVelMux::cmdVelCallback(const std::shared_ptr<geometry_msgs::msg::Twist> msg, unsigned int idx)
+bool CmdVelMux::addInputToParameterMap(std::map<std::string, ParameterValues> & parsed_parameters, const std::string & input_name, const std::string & input_variable, const rclcpp::Parameter & parameter_value)
```
@clalancette @Lobotuerk consider `input_variable` -> `parameter_name`

clalancette replied:
Fixed in 6091ab8532173b77448379318c2d7c1bb47dec2e

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -166,27 +221,120 @@ void CmdVelMux::configureFromParameters(const std::vector<std::string> & names,
+      continue;
+    }
+
+    if (!addInputToParameterMap(parsed_parameters, splits[0], splits[1], parameter.second))
```
@clalancette @Lobotuerk nit: aliases using const references may help readability

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -229,92 +377,62 @@ void CmdVelMux::timerCallback(unsigned int idx)
-    else
+
+    std::string input_name = splits[1];
+    std::string input_variable = splits[2];
```
@clalancette @Lobotuerk same as elsewhere, use const references where possible.

Lobotuerk committed:
Fixed and tested changes, adding and removing (#5)

* fixed priority dup, nit stuff, addition tested, add and remove

Lobotuerk commented on README.md:
```diff
@@ -15,8 +15,19 @@ priority or that stop publishing for any reason.  The stream currently in use is
+* `topic` (string) - The "topic" corresponding to each of the input `geometry_msgs/msg/Twist` topics.  This is what will be subscribed to.  The length of this list must match the `name` list above.
+* `timeout` (double) - The "timeout" of each of the input `geometry_msgs/msg/Twist` topics.  If no data is received on the input topic for this amount of time, the input will be automatically disabled.  The length of this list must match the `name` list above.
+* `priority` (integer) - The "priority" of each of the input `geometry_msgs/msg/Twist` topics.  The higher the number, the higher the priority.  Higher priority topics will dislodge lower priority topics and start publishing to the output topic automatically.  The length of this list must match the `name` list above.
+* `short_desc` (string) - The "short description" of each of the input `geometry_msgs/msg/Twist` topics.  This is informational only.  The length of this list must match the `name` list above.
```
Forgot to delete 'The length of this list must match the `name` list above.' in the description of the elements of the disctionary, sorry

clalancette replied:
No worries.  Feel free to open other PRs against this one, I'm happy to merge.

clalancette committed:
More fixes.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Get rid of duplication of topic, timeout, priority, and short_desc.

Juse re-use the ParameterValues struct.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Remove unnecessary commonTimerCallback().

I can't think of a situation in which the individual timers
will fail to remove themselves, so there is really no need
for the global timer.  Remove it here.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

Lobotuerk committed:
Added examples and changed the parameters explanation (#6)

clalancette committed:
More review fixes.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
I've integrated @Lobotuerk latest changes, finished a couple more of @hidmic's nits, and did some simplification of the code in here.  Another pass to confirm that what I've done seems reasonable would be appreciated.

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,144 +82,253 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
-    new_timeout = timeouts[i];
-    new_list[i]->priority_ = priorities[i];
-    new_list[i]->short_desc_ = short_descs[i];
+    new_map[key]->name_ = key;
```
@clalancette @Lobotuerk if `name_ == key` always holds `true`, what's the purpose of having `name_`?

Lobotuerk replied:
Explain pls

hidmic replied:
My point is that `CmdVelSub::name_` made sense back when we were not using an `std::map`. We have a map key, why storing a duplicate?

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,144 +82,253 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
-    new_list[i]->priority_ = priorities[i];
-    new_list[i]->short_desc_ = short_descs[i];
+    new_map[key]->name_ = key;
+    new_map[key]->values_.priority = parameter_values.priority;
```
@clalancette @Lobotuerk nit^2: you may also do `ParameterValues & new_parameter_values = new_map[key];` for improved readability below.

Lobotuerk replied:
Dont get this one either, could you explain further

hidmic replied:
I meant that instead of using `new_map[key]->values_` all over the place, we could do `ParameterValues & new_parameter_values = new_map[key]`. It's not shorter but it's easier to grok.

hidmic commented on src/cmd_vel_mux.cpp:
```diff
@@ -79,144 +82,253 @@ CmdVelMux::CmdVelMux(rclcpp::NodeOptions options) : rclcpp::Node("cmd_vel_mux",
+
+std::map<std::string, ParameterValues> CmdVelMux::parseFromParametersMap(const std::map<std::string, rclcpp::Parameter> & parameters)
+{
+  used_priorities_.clear();
```
@clalancette @Lobotuerk meta: I still think `used_priorities_` should not be a member field as it's tied to the `std::map` that this function populates and returns, but I don't mind merging as-is for the sake of moving forward.

clalancette replied:
Oh yeah, good point.  In point of fact, we have all of the infrastructure in place to just do this checking inside of `parametersAreValid`, so I moved it into there in 0ec9740c7fb63875e82d5af4228b8373e10c8259

hidmic reviewed (commented):
Overall looks good :)

clalancette committed:
Make used_priorities a local set.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

hidmic reviewed (approved):
LGTM, pending comments are minor

Lobotuerk commented on README.md:
````diff
@@ -12,11 +12,36 @@ priority or that stop publishing for any reason.  The stream currently in use is
+
+**Ex.1** for testing:
+```
+'ros2 service call /cmd_vel_mux/set_parameters_atomically rcl_interfaces/srv/SetParametersAtomically "{parameters: [{name : subscribers.new_input.topic, value: {type: 4, string_value: /input/new_input}}, {name : subscribers.new_input.priority, value: {type: 2, integer_value: 20}}, {name : subscribers.new_input.timeout, value: {type: 3, double_value: 0.2}}, {name : subscribers.new_input.short_desc ...
````
nit: Extra ' at the start of the line (my bad)

Lobotuerk committed:
Tested and fixed (#7)

* Fixes to map implementation for removal

Signed-off-by: lobotuerk <jtlorente@ekumenlabs.com>

clalancette commented:
With the latest fixes, this is in pretty good shape, and has been tested on a real robot.  I'm going to merge this to master now, and we can do any follow-ups later.
